### DISCUSSION TITLE
mixer shortcuts

### DISCUSSION THREAD

**Discussion by spamatica** [muse-sequencer/muse#1172] [Team Posts]
Hey Tim, I noticed you had worked on the shortcuts in the mixer.
Teaches me to work in the shadows, I also had a branch working on that :P  
...basically I'm a bit wary of my own ability to see a feature through, so I often don't want to announce stuff too early. In this case it backfired a bit as I was touching much the same things ;0)

Anyways, looking through your fixes I think I will attempt to make a merge of the two as I wanted to add a few more features (like adding the same shortcuts to the arranger). 

But - and here is my real question: Trying your changes I can't get the shortcuts to work. Ctrl+U to turn volume down for instance, does it work for you when a strip is selected?
Qt messing with us again? This is Kubuntu 18.04 and qt 5.9.5.
Also when selecting a strip it directly puts focus on the volume input field (which can be exited by using Esc), is it like that on your end too?
I can look at fixing/improving these things but I really want to know first if it behaves the same everywhere...

**Comment by spamatica** [comment]
Hey now. 
I got some of the shortcuts to work, WHILE the input field is active, but only then. Ctrl+U clears the input field instead.
I can go ahead and fix this, unless it works differently on your systems, guys?

**Comment by terminator356** [comment]
I bet those shortcuts are being eaten by whatever flavour of system you have.
Try changing the shortcuts in the shortcut dialog. (Now you know why I made them configurable.)
I know for example that Ctrl + A is reserved for selecting all text in a box.

However, once you do get the correct shortcuts, they should work always,
 no matter which control has current focus. Even if say, the instrument drop down has focus,
 those shortcuts are overriding and are supposed to always work.
They work fine here.

Do feel free to change those default keys once you are satisfied that they will work for everyone.
Do it quick before everyone pulls this as it is now, and the keys can't be changed easily by us.

Speaking of focus: Yes I carefully designed it that way.
Focus needs to 'land' on a sensible place to begin with.
Otherwise focus ends up in some no-mans-land place like the strip itself.
I found that I had to reinforce where it lands.
I followed my own general guidelines for 'smart focus' (there's a Setting for that)
 and when Escape is pressed, it 'gets you outta there' to a higher focus place.

**Comment by kybos** [comment]
> I got some of the shortcuts to work, WHILE the input field is active, but only then. Ctrl+U clears the input field instead.
> I can go ahead and fix this, unless it works differently on your systems, guys?

I checked that on Cinnamon - Ctrl+U is working, with and without focus in the input field.

**Comment by terminator356** [comment]
What happened to the mixer and strip focus scheme?

We are supposed to be able to place the focus on most of a strip's controls,
 and the control's text is therefore highlighted, so that for example pressing
 (shift and) up and down arrows will adjust the value up and down of a knob
 or horizontal slider (not the volume slider).

Also, tabbing was worked out so that almost every control could be 'visited' with tab,
 even from strip to strip on the mixer.

As it is now, the only control that seems to accept focus is the volume box.
What up with that?

**Comment by terminator356** [comment]
And the shortcuts to move between strips are gone from code but not shortcuts list?
Those are/were the Ctrl Home/End.
Accelerator keys are now back to being hard-coded in the mixer?
Configurable shortcuts for the mixer were to prevent that.

Several things seem to have changed in the code.

I mentioned that almost any control on a strip could gain focus before.
While it has focus, a control such as knob/slider will respond to up/down arrows (+ shift).
The strip volume and pan shortcuts **always** worked while the control had focus.
The strip found in the arranger and editors also responded to these keys.
We will not provide configurable shortcuts for the actual controls.
They have built-in keys like any other control.
So in the case of knob and slider, we must accept that up/down arrow are reserved for them.
So it was a matter of finding suitable strip-wide shortcuts, Ctrl Home/End seemed to work best
 for selecting, and the volume and pan, yeah that was tricky, only accel + letters seemed available.

**Comment by spamatica** [comment]
Ah, didn't digest all you wrote, I'll get back to you when time permits.
There were things I didn't get to work without some changes (and probably
some feature I did not grokk) but I wasn't supposed to leave any hardcoded
shortcuts

Den sön 28 juni 2020 08:41Tim <notifications@github.com> skrev:

**Comment by spamatica** [comment]
> What happened to the mixer and strip focus scheme?
Mistake on my part, I had changed the focus type on the internal widgets but it appears it was not necessary.

> And the shortcuts to move between strips are gone from code but not shortcuts list?
Fixed now

> I mentioned that almost any control on a strip could gain focus before.
> While it has focus, a control such as knob/slider will respond to up/down arrows (+ shift).
> The strip volume and pan shortcuts always worked while the control had focus.
> The strip found in the arranger and editors also responded to these keys.
> We will not provide configurable shortcuts for the actual controls.
> They have built-in keys like any other control.
> So in the case of knob and slider, we must accept that up/down arrow are reserved for them.
> So it was a matter of finding suitable strip-wide shortcuts, Ctrl Home/End seemed to work best
> for selecting, and the volume and pan, yeah that was tricky, only accel + letters seemed available.

I think I will go ahead with the release work even if there are still issues here.. even though I think I got it working but I'm still not entirely sure of the scope of this smart focus business.
I just wanted a quick way to change volume and pan in the mixer from the keyboard.

I see there is a checkbox in the settings for smart focus but no description on what it does, can one be written? or maybe it shouldn't be optional?

**Comment by terminator356** [comment]
Hi. There is a tooltip for that Smart Focus. Basically, when editing say, a box or spin, when return
 is pressed it will automatically return focus to the best place possible. For example, after editing
 say, the tempo box on the Arranger toolbar, pressing enter will return focus to the Part Canvas.
Without this Smart Focus the focus stays on the box, which is not good.
I made it optional in case it caused any problems, or users prefer to click focus on something else.

Escape does the exact same thing except it cancels any editing before yielding focus.

For the mixer strips it's a similar story.
After editing or escaping a box there, we can't just leave the focus lying around on the box,
 focus should be yielded to something higher, so that for example the transport shortcuts work.
Especially with escape. Pressing escape really should yield focus to something higher
 ie. "get me outta here".

That's what I was trying to do with all of this Smart Focus.
So you'll see that a several widgets or canvases have a 'focus yield' member for which focus
 will be given to if enter or escape is pressed.

**Comment by terminator356** [comment]
> I just wanted a quick way to change volume and pan in the mixer from the keyboard.

That was already in place. The only thing wrong was that the hard-coded shortcuts were not working.
That's why I made the shortcuts configurable.
Perhaps what happened was you saw that apparently there were no shortcuts at all
 but did not realize that there were, just not working?
Really the whole thing was worked out before (mostly by you) and just needed
 some existing shortcut TLC.

I hate to ask, but I urge you to consider backing up a step and undoing that commit.
It actually worked pretty well just before the commit, after your initial changes.
But this commit changed quite a few things and I fear it is a step backwards.
To fix what is wrong now will essentially put us back where we were before the commit.

**Comment by spamatica** [comment]
Den mån 13 juli 2020 00:35Tim <notifications@github.com> skrev:

> I just wanted a quick way to change volume and pan in the mixer from the
> keyboard.
>
> That was already in place. The only thing wrong was that the hard-coded
> shortcuts were not working.
> That's why I made the shortcuts configurable.
> Perhaps what happened was you saw that apparently there were no shortcuts
> at all
> but did not realize that there were, just not working?
> Really the whole thing was worked out before (mostly by you) and just
> needed
> some existing shortcut TLC.
>
> I hate to ask, but I urge you to consider backing up a step and undoing
> that commit.
> It actually worked pretty well just before the commit, after your initial
> changes.
> But this commit changed quite a few things and I fear it is a step
> backwards.
> To fix what is wrong now will essentially put us back where we were before
> the commit.
>
Are we still talking about the mixer? The tracklist had some bigger changes.

For the mixer I could probably revert the changes but could you give an
example what changed? I'm not really aware and thought what was left was
more cosmetic.

**Comment by spamatica** [comment]
I looked through the code one more time and apart from a a big chunk of commented code that I will remove, I'm still curious what might have stopped functioning. 
Basically I can't see that I have any functional changes left, just some rearrangements of the code, like a separate function for processing shortcuts (which was unnecessary as it turns out it only needs to be called from one place). 

Unless we are talking about the arranger/tracklist, but I don't think we are?

**Comment by terminator356** [comment]
One big thing that stopped functioning is the strips no longer accept focus
 on any controls except for volume. Each used to accept focus so that the up and down arrow
 could be used on any controls there. Then when enter or escape are pressed it returns focus
 to another place (the 'focus yield' widget).

**Comment by spamatica** [comment]
Ahhhh, I see... oh my, I never actually wrote that I had fixed that part, did I?! Just went straight into the comment that 'if there are issues I will ignore them, for now'...  I meant that in the context that I should have fixed the focus and the missing shortcuts.
Remaining changes in the mixer are to my knowledge only minor rearrangements of code. 
Still kinda struggling to find time, hopefully I can do the release work on friday.

**Comment by terminator356** [comment]
Status: I got called back to work. And I've been very quiet here because I've been doing
 some critical database work. (I wrote the software that my old shop used for 10 years.)
Business database work can be lucrative since it's crucial for operations and many businesses
 are reluctant to switch to another product - or even another computer. Software that they purchased
 with a license key may only run on that computer since the software company may no longer exist.
So it's crucial to maintain the computer and the software. We fix 20 year old P4 PCs with bad caps
 instead of replacing them for those very reasons.
For example 16-bit Point-Of-Sale terminal software got bit by Windows 64-bit - no 16-bit subsystem.
Hello saviour DosBox !

I mentioned I got an Android tablet, so I've also been playing with that as well and... attempting to
 build MusE, but no luck even getting the dev chain set up for now. It's confusing...

Meanwhile I'm slowly working on the MusE midi looping as requested. It's coming along.
I'll be happy when that's over and out of the way.

**Comment by spamatica** [comment]
Den ons 15 juli 2020 kl 01:39 skrev Tim <notifications@github.com>:

> Status: I got called back to work. And I've been very quiet here because
> I've been doing
> some critical database work. (I wrote the software that my old shop used
> for 10 years.)
> Business database work can be lucrative since it's crucial for operations
> and many businesses
> are reluctant to switch to another product - or even another computer.
> Software that they purchased
> with a license key may only run on that computer since the software
> company may no longer exist.
> So it's crucial to maintain the computer and the software. We fix 20 year
> old P4 PCs with bad caps
> instead of replacing them for those very reasons.
> For example 16-bit Point-Of-Sale terminal software got bit by Windows
> 64-bit - no 16-bit subsystem.
> Hello saviour DosBox !
>
Oh wow, never done any of that kind of work but I can totally see it
happening over and over.
Tried moving the business stuff code to a VM, to disconnect it from the
archaic hardware? Ah probably the license key will fuck that up...

Well, work is work, hope you get through it without losing to much sanity.

> I mentioned I got an Android tablet, so I've also been playing with that
> as well and... attempting to
> build MusE, but no luck even getting the dev chain set up for now. It's
> confusing...
>
Hehe., yeah.. Android may be linux underneath but getting access to that is
no walk in the park... if it works at all.


> Meanwhile I'm slowly working on the MusE midi looping as requested. It's
> coming along.
> I'll be happy when that's over and out of the way.
>
Oh, no worries, we're supposed to have summer break, right? ;)
Midi looping though, drawing a blank there, is it the punch in/out stuff?

---
Started on the release work, unless something happens it'll be up tomorrow.

Other than that I'm trying to finally get a bit of vacation. Actually it's
the start of something completely different as I will be taking a very long
break from work and go to school after the summer holidays... \o/.. or,
we'll see.
